src/VegaWrapper2.js

```diff
@@ -0,0 +1,403 @@
+        } else if (this.testHost('http', host)) {
+            return { host: host, protocol: 'http' };
+        }
+        return undefined;
+    };
+
+    /**
+     * Test host against the list of allowed domains based on the protocol
+     * @param {string} protocol
+     * @param {string} host
+     * @returns {boolean}
+     */
+    testHost(protocol, host) {
+        if (!this.validators[protocol]) {
+            const domains = this.domains[protocol];
+            if (domains) {
+                this.validators[protocol] = makeValidator(domains, protocol === 'https' || protocol === 'http');
+            } else {
+                return false;
+            }
+        }
+        return this.validators[protocol].test(host);
+    };
+
+    /**
+     * Override the protocol and host for *wikidatasparql*, *geoshape*, *geoline* and *mapsnapshot*
+     * @param {object} urlParts output the new result to this object
+     * @param {object} urlObj used to log and determine the corresponding host
+     * @param {string} protocolOverride used to determine the corresponding host
+     * @private
+     */
+    _overrideHostAndProtocol(urlParts, urlObj, protocolOverride) {
+        let protocol = protocolOverride || urlObj.type,
+            domains = this.domains[protocol];
+        if (!domains) {
+            throw new Error(protocol + ': protocol is disabled: ' + JSON.stringify(urlObj));
+        }
+        urlParts.host = domains[0];
+        urlParts.protocol = this.sanitizeHost(urlParts.host).protocol;
+    }
+
+    /**
+     * convert the urlObj to a url string
+     * @param {object} urlObj an object consists of type and essential parameters
+     * @param {object} options used to attach CORS infomation
+     * @returns {string} a complete url
+     */
+    objToUrl(urlObj, options) {
+        const sanitizedHost = this.sanitizeHost(urlObj.wiki ? urlObj.wiki : options.domain);
+
+        if (!sanitizedHost) {
+            throw new Error('URL hostname is not whitelisted: ' + urlParts.host);
+        }
+        const urlParts = {
+            host: sanitizedHost.host,
+            protocol: sanitizedHost.protocol,
+            query: {}
+        };
+
+        switch(urlObj.type) {
+            case 'wikiapi':
+                // {type: “wikiapi”, params: {action:”...”, ...} [, wiki: “en.wikipedia.org”]}
+                // Call to api.php - the *params* are converted into the url query string
+                // use *wiki* to designate the host
+                if(!urlObj.params || typeof urlObj.params !== 'object') {
+                    throw new Error('wikiapi: "params" should be an object');
+                }
+                if(Array.isArray(urlObj.params)) {
+                    throw new Error('wikiapi: "params" shouldn\'t be an array');
+                }
+                for(const k of Object.keys(urlObj.params)) {
+                    const v = urlObj.params[k];
+                    if(typeof v === 'object') {
```

Comment:
otherwise you will allow `undefined` and `function`, as well as some other possible values in the future (who knows what JS might introduce)
```suggestion
                    if(typeof v !== 'number' && typeof v !== 'string' && typeof v !== 'boolean') {
```

Reply:
Actually, I don't think `undefined` or `function` can be passed to here from a JSON definition.
But you are right, we don't know the future.